### ISSUE TITLE
[Question] How to yield chunks of a zipped file

### ISSUE THREAD

**Issue by alexanderluiscampino** [uktrade/stream-unzip#26]
Hi,

I am trying to create a low-memory stream unzipper that can be ran in AWS Lambda. Currently I can unzip individual files from a .zip package. But the problem I am facing is that I need to read that one file into memory, which for a large file, 15gb, it is being killed by the runtime. 

I came across this library, but I am unsure it delivers what I need. Or maybe it does but I don't see how.  

This is my example below, using the normal `zipfile` from the standard lib.

```python
import zipfile 
import s3fs
_zipfile_interface = zipfile.ZipFile
_s3_interface = s3fs.S3FileSystem(anon=False)
_READING_MODE = 'rb'
_WRITING_MODE = 'wb'


def extract_file(filename: str, output_uri: str):
    """
    Extracts filename from zip package into input output S3 URI

    Args:
        filename (str): Name of the file inside zip package
        output_uri (str): Output S3 URI

    Raises:
        exceptions.ExtractionFailure: In case extraction errors out
    """
    with _s3_interface.open(s3_uri, _READING_MODE) as f_in:
        zip_buffer = _zipfile_interface(f_in)

        file_buffer = zip_buffer.open(filename)

        with _s3_interface.open(output_uri, _WRITING_MODE) as f_out:

            f_out.write(file_buffer.read())  # -> the issue lies here
```

Do ignore S3 interface, for purpose of this example, it's just a normal `open()`.

I have profiled this method here:
```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    78     66.2 MiB     66.2 MiB           1       @profile
    79                                             def extract_file(filename: str, output_uri: str):
    80                                                 """
    81                                                 Extracts filename from zip package into input output S3 URI
    82                                         
    83                                                 Args:
    84                                                     filename (str): Name of the file inside zip package
    85                                                     output_uri (str): Output S3 URI
    86                                         
    87                                                 Raises:
    88                                                     exceptions.ExtractionFailure: In case extraction errors out
    89                                                 """
    90     66.2 MiB      0.0 MiB           1           try:
    91     79.1 MiB      2.1 MiB           2               with _s3_interface.open(package.s3_uri, _READING_MODE) as f_in:
    92     73.1 MiB      0.0 MiB           1                   zip_buffer = _zipfile_interface(f_in)  # type: ignore
    93                                         
    94     83.9 MiB     10.8 MiB           1                   file_buffer = zip_buffer.open(filename)
    95                                         
    96     83.9 MiB      0.0 MiB           2                   with _s3_interface.open(output_uri, _WRITING_MODE) as f_out:
    97                                         
    98     83.9 MiB      0.0 MiB           1                       f_out.write(file_buffer.read())  # type: ignore
    99                                         
```

Line 94 you see the jump in memory, 10MB, from the `zip.open(filename)` this is when the file is consumed into memory and then I can write it out to the output buffer. 

Ideally, I would like that `zip.open` to iterate chunks of size N that I can just write to the open buffer. 

Any knowledge on this?

Thank you,
Alex

**Comment by michalc** [comment]
You should be able to avoid loading both the zip and any member files into memory - it's the main reason for the existence of stream-unzip!

Something like the below should work:

```python
import os
import s3fs
from stream_unzip import stream_unzip

def extract_file(fs, input_uri, input_member_file_name, output_uri):

    with \
            fs.open(input_uri, 'rb') as f_in, \
            fs.open(output_uri, 'wb') as f_out:

        for file_name, file_size, unzipped_chunks in stream_unzip(f_in):
            # Skip any files in zip that aren't needed
            if file_name != input_member_file_name.encode():
                for _ in unzipped_chunks:
                    pass
                continue

            for chunk in unzipped_chunks:
                f_out.write(chunk)

extract_file(
    fs=s3fs.S3FileSystem(anon=False),
    input_uri='my-bucket/my-input-file.zip',
    input_member_file_name='my-member-file.txt',
    output_uri='my-bucket/my-output-file.txt',
)
```

**Comment by alexanderluiscampino** [comment]
@michalc You're right, that does the trick, I got to similar solution after banging my head late at night. The issue was time. 
This part of the code:
```python
if file_name != input_member_file_name.encode():
                for _ in unzipped_chunks:
                    pass
                continue
```
The chunks need to be iterated through either I want them or not. That involves many calls to S3, which some got throttled and the whole thing blew up. Or, it took too long to get to my file of interest... ran for over one hour. 

I do have a very complex scenario, where I get this huge .zip packages (>10GB), 100s of files. Then files inside it might get to 30GB uncompressed. With 100% compression rate, which makes the machine work extra hard to unzip. 

In short, stream-unzip with the solution above does what it promises! Unfortunately, does not resolve my issue within my constraints :/

**Comment by michalc** [comment]
> The chunks need to be iterated through either I want them or not

Yes this is true - there was discussion on how this can be surprising at https://github.com/uktrade/stream-unzip/issues/21, but so far I'm still of the opinion that stream-unzip shoul require this

> In short, stream-unzip with the solution above does what it promises! Unfortunately, does not resolve my issue within my constraints :/

Ah a shame - but in that case, I'll close this issue (but feel free to re-open if I've misunderstood, or open another one)
